src/index/refine_flat.rs

```diff
@@ -80,6 +80,25 @@ impl<BI> FromInnerPtr for RefineFlatIndexImpl<BI> {
     }
 }
 
+impl<BI> TryFromInnerPtr for RefineFlatIndexImpl<BI> {
```

Comment:
Hmm, we may have a small soundness issue with `TryFromInnerPtr`. The cast can validate that the index is indeed a `RefineFlatIndex`, but it cannot validate the type parameter `BI`. I don't see that we can unwrap the index into one of type `BI` (which could be awfully dangerous), but `CpuIndex` and `Send` + `Sync` are conditionally implemented on whatever `BI` is, so this could easily fall off the memory safety boundaries in a way that cannot be checked at run-time.

Both methods in `*FromInnerPtr` are still unsafe though, so we can at least leave the hint that the type parameter is _not validated_, either here or in the two traits.

Reply:
I agree. I will think how it to fix.